cpp/dolfinx/io/ADIOS2Writers.h

```diff
@@ -830,6 +830,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
 }
 } // namespace impl_vtx
 
+/// @privatesection
+namespace impl_checkpoint
+{
+/// Write mesh to file using DOLFINx mesh layout.
+/// @param[in] io ADIOS2 io object
+/// @param[in] engine ADIOS2 engine
+/// @param[in] mesh Mesh to write.
+template <typename T>
+void write_mesh(adios2::IO& io, adios2::Engine& engine,
+                const mesh::Mesh<T>& mesh)
+{
+  const mesh::Geometry<T>& geometry = mesh.geometry();
+
+  {
+    // "Put" geometry
+    std::size_t num_xdofs_local = geometry.index_map()->size_local();
+    std::int64_t num_xdofs_global = geometry.index_map()->size_global();
+    std::array<std::int64_t, 2> local_range
+        = geometry.index_map()->local_range();
+
+    std::span<const T> x = geometry.x();
+    std::size_t gdim = geometry.dim();
+
+    // Trim geometry to gdim and trunctate for owned entities only
+    std::vector<T> buffer(num_xdofs_local * gdim);
+    for (std::size_t i = 0; i < num_xdofs_local; ++i)
+      for (std::size_t j = 0; j < gdim; ++j)
+        buffer[i * gdim + j] = x[3 * i + j];
+
+    // Define geometry variable and "put"
+    adios2::Variable<T> local_geometry = impl_adios2::define_variable<T>(
+        io, "geometry", {num_xdofs_global, gdim}, {local_range[0], 0},
+        {num_xdofs_local, gdim});
+    engine.Put<T>(local_geometry, buffer.data());
+  }
+
+  // Extract topology (cell) information
+  auto topology = mesh.topology();
+  assert(topology);
+  auto cell_imap = topology->index_map(topology->dim());
+  const std::size_t num_cells_local = cell_imap->size_local();
+  const std::size_t num_cells_global = cell_imap->size_global();
+  const std::size_t local_range_cell = cell_imap->local_range()[0];
+
+  // Extract geometry information
+  if (geometry.cmaps().size() > 1)
+    throw std::runtime_error("Multiple coordinate maps not supported");
+  fem::ElementDofLayout geom_layout = geometry.cmaps()[0].create_dof_layout();
+  std::size_t num_dofs_per_cell
+      = geom_layout.num_entity_closure_dofs(geometry.dim());
+  auto geometry_imap = geometry.index_map();
+  auto geometry_dmap = geometry.dofmap();
+
+  // Convert local geometry to global geometry
+  std::vector<std::int64_t> connectivity_out(num_cells_local
```

Comment:
The terminology here is confusing. The variable name includes 'connectivity', but that data that gets copied in is geometry data.